### ISSUE TITLE
Dynamic routing for cmdroute.Router

### ISSUE THREAD

**Issue by FlowingSPDG** [diamondburned/arikawa#438]
Hi again!

I am developing tournament Discord BOT, and I madea BOT that does following:

1. User creates a tournament with slash command
2. BOT made a tournament, got Tournament ID in our Database. Respond "JOIN THIS TOURNAMENT!" button as a interaction.
3. Player hits the button, but we don't have the tournament ID so we can't process from here

I was trying to figure out somehow, but found out interaction customID needs be flexible e.g.``join_tournament/(tournamentID)``.
currently cmdRouter solves it's node by component custom ID.
https://github.com/diamondburned/arikawa/blob/v3/api/cmdroute/router.go#L216-L220
https://github.com/diamondburned/arikawa/blob/v3/api/cmdroute/router.go#L404-L410

that means ``(*cmdroute.Router).AddComponentFunc("join_tournament/(tournamentID)"`` can't be solved.

So my idea is implement flexible path handler like gin-gonic/gin:

https://github.com/gin-gonic/examples/blob/master/realtime-chat/main.go#L18-L22

```go
func (r *Router) handleComponent(ev *discord.InteractionEvent, component discord.ComponentInteraction) *api.InteractionResponse {
	// Get path parameter here and handle with it
	// e.g. 
	// (*Router).AddComponentFunc("join_tournament/:tournament_id")
	id := component.ID()
	fmt.Println("ID: ", id) // join_tournament/1

	params := r.parseParams(id) 
	baseCommand := params.Base()
	fmt.Println("base command: ", baseCommand) // join_tournament
	fmt.Println("tournament ID: ", params.Param("tournament_id")) // 1

	node, ok := r.nodes[baseCommand ].(routeNodeComponent)
	if ok {
		return r.callComponentHandler(ev, node.component, params)
	}
	return nil
}
```

**Comment by diamondburned** [comment]
You can implement your own slug handler like so [[playground](https://go.dev/play/p/fuvfR7mvD4V)]:

```go
package main

import (
	"context"
	"log"
	"strings"

	"github.com/diamondburned/arikawa/v3/api"
	"github.com/diamondburned/arikawa/v3/api/cmdroute"
	"github.com/diamondburned/arikawa/v3/api/webhook"
	"github.com/diamondburned/arikawa/v3/discord"
	"github.com/diamondburned/arikawa/v3/gateway"
	"github.com/diamondburned/arikawa/v3/session"
)

func main() {
	ctx := context.Background()

	r := cmdroute.NewRouter()
	r.Add("command", nil)

	s := session.New("").WithContext(ctx)
	s.AddInteractionHandler(r)
	s.AddInteractionHandler(handleSlugs(ctx, map[string]cmdroute.InteractionHandler{
		"join_tournament": cmdroute.InteractionHandlerFunc(handleJoinTournament),
	}))

	s.Handler.Call(&gateway.InteractionCreateEvent{
		InteractionEvent: discord.InteractionEvent{
			Data: &discord.ButtonInteraction{
				CustomID: "join_tournament/aud92j3",
			},
		},
	})

	<-joinedTournament
}

var joinedTournament = make(chan struct{}, 1)

func handleJoinTournament(ctx context.Context, ev *discord.InteractionEvent) *api.InteractionResponse {
	tournamentID := string(slugIDFromContext(ctx))
	log.Println("joined tournament", tournamentID)
	joinedTournament <- struct{}{}
	return nil
}

type slugID string

func slugIDFromContext(ctx context.Context) slugID {
	v, _ := ctx.Value(slugID("")).(slugID)
	return v
}

func handleSlugs(ctx context.Context, handlerPrefixes map[string]cmdroute.InteractionHandler) webhook.InteractionHandler {
	return webhook.InteractionHandlerFunc(func(ev *discord.InteractionEvent) *api.InteractionResponse {
		componentInteraction, ok := ev.Data.(discord.ComponentInteraction)
		if !ok {
			return nil
		}

		prefix, arg, ok := strings.Cut(string(componentInteraction.ID()), "/")
		if !ok {
			return nil
		}

		handler, ok := handlerPrefixes[prefix]
		if !ok {
			return nil
		}

		slug := slugID(arg)
		ctx = context.WithValue(ctx, slugID(""), slug)
		return handler.HandleInteraction(ctx, ev)
	})
}
```

**Comment by diamondburned** [comment]
I'm personally against adding a slug parser into `cmdroute` since it'll make the code a lot more complicated than it already is. `cmdroute` is supposed to be a simple router that you can easily extend, and combining that with a lot of advanced features is really hard. 

I'm closing this issue for this reason, but your use case should still be doable with `cmdroute` as it is right now.

**Comment by FlowingSPDG** [comment]
@diamondburned  I got your idea, thanks! I respect that 👍 

One reason I wanted to add that is most of my simple commands(w/o arguments) can be done with cmdRouter and ``join_tournament`` and some other commands cant, and I don't want to mix up two-routers(my router and cmdRouter) for maintainability so I decided to not use cmdRouter.

maybe adding more "complex" router based on cmdRouter is more reasonable? e.g. typed datas with pre-assertion, slug handler etc?
